### PR TITLE
Add a concept of "Test Mode"

### PR DESCRIPTION
**Pull request by superm1**
Rather than plugins being disabled via the `DisabledPlugins` key
treat the `test` and `test_ble` plugins as special plugins that
are only enabled in test mode.

When the `TestMode` key is set in `/etc/fwupd/fwupd.conf` then these
plugins will be allowed to load.

Furthermore there is a remote loaded on all systems called "fwupd-tests"
which is only used for the test suite.

When `TestMode` key is set then it will be allowed to be used.

Also introduce and use helper args for fwupdtool to turn on/off test
mode.  These helper args are:

`fwupdtool enable-test-mode` and `fwupdtool disable-test-mode`.

When called, 3 things will happen:
    * Allow using the fwupd-tests remote
    * Allow the test plugins to load
    * Allow emulation by default (overrides `AllowEmulation` key)

With all of these changes in place fwupd.conf can be populated
ENTIRELY EMPTY!

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: data/fwupd.conf [hunk 1]
```diff
@@ -1,5 +1,2 @@
-# use `man 5 fwupd.conf` for documentation
 [fwupd]
-DisabledPlugins=test;test_ble
-OnlyTrusted=true
-AllowEmulation=false
+# use `man 5 fwupd.conf` for documentation
```
Comment (superm1) [inline data/fwupd.conf:2]: Reason for this change is the key file parser doesn't like a totally empty section.... 😄

File: src/fu-tool.c [hunk 2]
```diff
@@ -2052,6 +2052,43 @@ fu_util_remote_enable(FuUtilPrivate *priv, gchar **values, GError **error)
...
+	remote = fu_engine_get_remote_by_id(priv->engine, "fwupd-tests", error);
+	if (remote == NULL)
+		return FALSE;
+
+	if (!fu_engine_modify_remote(priv->engine,
+				     fwupd_remote_get_id(remote),
+				     "Enabled",
+				     enable ? "true" : "false",
+				     error))
+		return FALSE;
```
Comment (superm1) [inline src/fu-tool.c:2072]: the other idea I have for this is to special case the `fwupd-tests` remote where it's only loaded if the engine is in test mode.  Then it can default to `Enabled` in the conffile but _normally_ not be used.

File: .github/workflows/snap.yml [hunk 3]
```diff
@@ -46,7 +46,7 @@ jobs:
       run: |
         sudo snap install --dangerous ${{ needs.build-snap.outputs.snap_name }}
         sudo snap connect fwupd:polkit :polkit
-        sudo fwupd.fwupdtool modify-config DisabledPlugins ""
+        sudo fwupd.fwupdtool enable-test-mode
```
Comment (hughsie) [inline .github/workflows/snap.yml:49]: I like the concept, I just don't like the *name*. Mode implies one or the other, and this kinda just enables an extra optional part of fwupd.

File: src/fu-engine.c [hunk 4]
```diff
@@ -6452,6 +6453,10 @@ fu_engine_is_plugin_name_disabled(FuEngine *self, const gchar *name)
 		if (g_strcmp0(name_tmp, name) == 0)
 			return TRUE;
 	}
+
+	if (!fu_engine_config_get_test_mode(self->config) && g_str_has_prefix(name, "test"))
```
Comment (hughsie) [inline src/fu-engine.c:6457]: this seems super fragile. Could we have those two plugins adding `FWUPD_REMOTE_FLAG_ONLY_TEST_MODE` in their `init()` and then checking for that instead?

File: .github/workflows/snap.yml (hunk 3 above)
Comment (superm1) [reply .github/workflows/snap.yml:49]: Got another preference?  Some ideas:
`*able-test-infrastructure`
`*able-test-devices`
`*able-virtual-devices`
`*able-virtual-test-devices`

Comment (hughsie) [reply .github/workflows/snap.yml:49]: I guess the name needs to reflect what it does. It's really providing another couple of plugins that export new virtual devices we can play with. Maybe `enable-test-devices` is best?

File: src/fu-engine.c (hunk 4 above)
Comment (superm1) [reply src/fu-engine.c:6457]: Yeah; I'll work out a new flag for them to use instead.

File: .github/workflows/snap.yml (hunk 3 above)
Comment (superm1) [reply .github/workflows/snap.yml:49]: Or even `*able-testing` or `*able-test-suite`

File: src/fu-engine.c (hunk 4 above)
Comment (superm1) [reply src/fu-engine.c]: fixed

File: .github/workflows/snap.yml (hunk 3 above)
Comment (superm1) [reply .github/workflows/snap.yml]: fixed

File: src/fu-engine-config.c [hunk 5]
```diff
@@ -338,9 +339,23 @@ fu_engine_config_get_show_device_private(FuEngineConfig *self)
...
+					FU_DAEMON_CONFIG_DEFAULT_TEST_DEVICES);
+}
+
 gboolean
 fu_engine_config_get_allow_emulation(FuEngineConfig *self)
 {
+	if (fu_engine_config_get_test_devices(self)) {
+		g_info("allowing emulation as TestDevices is enabled");
+		return TRUE;
+	}
```
Comment (hughsie) [inline src/fu-engine-config.c:357]: on the logic? seems a bit orthogonal, no?

File: docs/fwupd.conf.md [hunk 6]
```diff
@@ -202,6 +202,11 @@ The `[fwupd]` section can contain the following parameters:
 
   At some point in the future fwupd will change the default to `metadata,firmware`.
 
+**TestDevices={{FU_DAEMON_CONFIG_DEFAULT_TEST_DEVICES}}**
+
+  Chich creates virtual test devices and remote for validating daemon flows.
```
[the remark this answers is not in the document]
Comment (superm1) [reply docs/fwupd.conf.md:207]: Typo :)

File: src/fu-tool.c [hunk 7]
```diff
@@ -2052,6 +2052,34 @@ fu_util_remote_enable(FuUtilPrivate *priv, gchar **values, GError **error)
...
+	if (!fu_util_start_engine(priv, FU_ENGINE_LOAD_FLAG_NONE, priv->progress, error))
+		return FALSE;
+
+	if (!fu_engine_modify_config(priv->engine, "TestDevices", enable ? "true" : "false", error))
+		return FALSE;
+
+	if (enable)
+		fu_console_print_literal(priv->console, _("Successfully enabled test devices"));
+	else
+		fu_console_print_literal(priv->console, _("Successfully disabled test devices"));
```
Comment (hughsie) [inline src/fu-tool.c:2067]: needs TRANSLATOR context -- and braces

File: src/fu-engine-config.c (hunk 5 above)
Comment (superm1) [reply src/fu-engine-config.c:357]: Well the fwupd.sh tests run all sorts of emulation device tests, which seemed to fit the bill for "virtual devices"

Comment (hughsie) [reply src/fu-engine-config.c:357]: Seems like a different thing to me.

Comment (superm1) [reply src/fu-engine-config.c:357]: If you really think they shouldn't be linked; then I think fwupd.sh needs `fwupdtool modify-config AllowEmulation true`

Comment (hughsie) [reply src/fu-engine-config.c]: I think it's confusing to link test-devices and emulation -- they're two completely different things, and enabling the former isn't obvious that the latter is also enabled.

Comment (superm1) [reply src/fu-engine-config.c]: OK; look at the fixup commit.  Does that make more sense to you?  If that passes CI I'll squash it.

File: data/installed-tests/fwupd.sh [hunk 8]
```diff
@@ -13,6 +13,8 @@ run_test()
 run_device_tests()
 {
 	if [ -n "$CI_NETWORK" ] && [ -d @devicetestdir@ ]; then
+		systemctl stop fwupd.service
+		fwupdtool modify-config AllowEmulation true
```
Comment (hughsie) [inline data/installed-tests/fwupd.sh:17]: if you used fwupd**mgr** modify-config AllowEmulation true could you avoid stopping the service?

Comment (superm1) [reply data/installed-tests/fwupd.sh:17]: I'll try.

Comment (superm1) [reply data/installed-tests/fwupd.sh]: Got it working.

**Review by hughsie — APPROVED** [review]
